tokio/docs/runtime-stats.md

```diff
@@ -0,0 +1,137 @@
+- Blocking task queue depth
+
+### Task
+
+This implementation will avoid tracking stats/metrics at the task level due to the overhead required. This will instead be accomplished by the [tokio console](https://github.com/tokio-rs/console). This will allow the user to attach the console and take the performance hit when they want to explore issues in more detail.
+
+### I/O driver
+
+Unlike, the executor stats, stats coming from the I/O driver will be streamed directly to the `Stats` struct via atomics. Each value will be incremented (via `AtomicU64::fetch_add`) for each event.
+
+List of stats provided from the I/O driver:
+- Amount of compact
+- Amount of "token dispatches" (aka ready events)
+- Amount of fd currently registered with `io::Driver`
+- Amount of read ready events
+- Amount of write ready events
+
+### Timer driver
+
+Similar to the I/O driver the timer driver will stream stats in a similar fashion. These will be emitted via `AtomicU64::fetch_add` and be emitted on each occurance of the events below.
+
+List of stats provided from the timer driver:
+- Amount of timer entries (timer wheel depth)
+- Amount of timers fired
+- Amount of timers cancelled
+
+### `tokio-metrics`
+
+The `tokio-metrics` crate will provide aggregated metrics based on the `Stats` struct. This will include histograms and other useful aggregated forms of the stats that could be emitted by various metrics implementations. This crate is designed to provide the ability to expose the aggregated stats in an unstable `0.1` method outside of the runtime and allow the ability to iterate on how they are aggregated without the need to follow `tokio`'s strict versioning scheme.
+
+The rest of this crate is out of scope for this RFC, since most of the implementation will happen ad-hoc outside of the main crate.
+
+## Implementation details
+
+### Batch streaming
+
+To avoid any extra overhead in the executor loop, each worker will batch metrics into a `Core` local struct. These values will be incremented or sampled during regular executor cycles when certain operations happen like a work steal attempt or a pop from one of the queues.
+
+The batches will be streamed via atomics to the stats struct directly. This will reduce any cross CPU work while the executor is running and amortize the cost of having to do cross CPU work. Batches will be sent before the executor attempts to park the thread. This will happen either when there is no work to be done or when the executor has hit the maintance tick. At this point before the thread will park, the executor will submit the batch. Generally, since parking is more expensive then submitting batches there should not be any added latency to the executor cycle in this process.
+
+This then allows the collector to poll the stats on any interval. Thus, allowing it to drive its own timer to understand an estimate of the duration that a certain amount of ticks too, or how many times the the executor entered the park state.
```

Comment:
nit: structure/flow

This sentence has grammatical issues, but I think the bigger problem is that it potentially conflicts with the idea that the executor ticks on a predictable interval.  Why would we need to track the duration vs ticks ratio ourselves?